std/assembly/string.ts

```diff
@@ -342,7 +341,123 @@ import { idof } from "./builtins";
+          (len - index) << 1
+        );
+        return changetype<String>(out);
+      }
+    }
+    return this;
+  }
+
+  replaceAll(search: String, replacement: String): String {
+    assert(this !== null && search !== null && replacement !== null);
+    var len: usize = this.length;
+    var slen: usize = search.length;
+    if (len <= slen) {
+      return len < slen ? this : select<String>(replacement, this, search == this);
+    }
+    var rlen: usize = replacement.length;
+    if (!slen) {
+      if (!rlen) return this;
+      // Special case: 'abc'.replaceAll('', '-') -> '-a-b-c-'
+      let out = __alloc((len + (len + 1) * rlen) << 1, idof<String>());
+      memory.copy(out, changetype<usize>(replacement), rlen << 1);
+      let offset = rlen;
+      for (let i: usize = 0; i < len; ++i) {
+        store<u16>(
+          changetype<usize>(out) + (offset++ << 1),
+          load<u16>(changetype<usize>(this) + (i << 1))
+        );
+        memory.copy(
+          out + (offset << 1),
+          changetype<usize>(replacement),
+          rlen << 1
+        );
+        offset += rlen;
+      }
+      return changetype<String>(out);
+    }
+    var prev: isize = 0, next: isize = 0;
+    if (slen == rlen) {
+      // Fast path when search and replacement have same length
+      let size = len << 1;
+      let out = __alloc(size, idof<String>());
+      memory.copy(out, changetype<usize>(this), size);
+      while (~(next = <isize>this.indexOf(search, <i32>prev))) {
+        memory.copy(out + (next << 1), changetype<usize>(replacement), rlen << 1);
+        prev = next + slen;
+      }
+      return changetype<String>(out);
+    }
+    var out: usize = 0, offset: usize = 0, resLen = len;
+    while (~(next = <isize>this.indexOf(search, <i32>prev))) {
+      if (!out) out = __alloc(len << 1, idof<String>());
+      if (offset > resLen) {
+        let newLength = resLen << 1;
+        out = __realloc(out, newLength << 1);
+        resLen = newLength;
+      }
+      let chunk = next - prev;
+      memory.copy(
+        out + (offset << 1),
+        changetype<usize>(this) + (prev << 1),
+        chunk << 1
+      );
+      offset += chunk;
+      memory.copy(
+        out + (offset << 1),
+        changetype<usize>(replacement),
+        rlen << 1
+      );
+      offset += rlen;
+      prev = next + slen;
+    }
+    if (offset) {
+      if (offset > resLen) {
+        let newLength = resLen << 1;
+        out = __realloc(out, newLength << 1);
+        resLen = newLength;
+      }
+      let rest = len - prev;
+      if (rest) {
+        memory.copy(
+          out + (offset << 1),
+          changetype<usize>(this) + (prev << 1),
+          rest << 1
+        );
+      }
+      rest += offset;
+      if (resLen > rest) out = __realloc(out, rest << 1);
+      return changetype<String>(out);
+    }
+    return this;
+  }
+
   slice(start: i32, end: i32 = i32.MAX_VALUE): String {
+    assert(this !== null);
```

Comment:
Left over :)

Reply:
Should I remove `assert(this !== null)` in all rest places?

Reply:
Yeah, should probably remove all of these. There have been a few occasions where a parameter type should actually be `string | null` as well. Fixed some of these with the runtime pr, but it's possible that there are others remaining.

Reply:
done!